File: 059_SpiralMatrix_II.py
def generateMatrix(self, n):
    A = [[0] * n for _ in range(n)]
    i, j, di, dj = 0, 0, 0, 1
    for k in xrange(n*n):
        A[i][j] = k + 1
        if A[(i+di)%n][(j+dj)%n]:
            di, dj = dj, -di
        i += di
        j += dj
    return A
# Can not work
class Solution(object):
    def generateMatrix(self, n):
        """
        :type n: int
        :rtype: List[List[int]]
        """
        if n <= 0:
            return list()
        subRes = range(n)
        res = [subRes for i in range(n)]
        up, down, left, right = 0, n - 1, 0, n - 1
        index = 1
        while True:
            #Up
            for col in xrange(left, right + 1):
                res[up][col] = index
                index += 1
            up += 1
            if up > down: break
            
            #Right
            for row in xrange(up, down + 1):
                res[row][right] = index
                index += 1
            right -= 1
            if right < left: break
            
            #Down
            for col in (range(left, right + 1))[::-1]:
                res[down][col] = index
                index += 1
            down -= 1
            if down < up: break
            
            #Left
            for row in (range(up, down + 1))[::-1]:
                res[row][left] = index
                index += 1
            left += 1
            if left > right: break
        return res
        
        

File: 059_SpiralMatrix_II.cpp
//33.14%
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
		if(n <= 0){
			vector<vector<int>> temp;
			return temp;
		}
		
        vector<int> vec(n);
		vector<vector<int>> res(n, vec);
		int up(0), down(n - 1), left(0), right(n - 1);
		int index(1);
		while(true){
			//up
			for(int col = left; col <= right; col++)
				res[up][col] = index++;
			if(++up > down) break;
			//right
			for(int row = up; row <= down; row++)
				res[row][right] = index++;
			if(--right < left) break;
			//down
			for(int col = right; col >= left; col--)
				res[down][col] = index++;
			if(--down < up) break;
			//left
			for(int row = down; row >= up; row--)
				res[row][left] = index++;
			if(++left > right) break;
		}
		return res;
    }
};